### DISCUSSION TITLE
出码如何实现即时刷新

### DISCUSSION THREAD

**Discussion by stwflyfox** [opentiny/tiny-engine#449] [General]
我想要实现的是出码的页面跟项目集成是保存到数据库中，而不是把页面打包到项目里，在打开页面时是从数据库中读取显示的。这样在设计器对页面做了修改时把出码更新到数据库，不用重新打包，只要重新打开页面就能即时刷新。另外已打开的页面要能KeepALive 。 是否有好的解决方案？

**Comment by chilingling** [comment]
1. 出码的页面集成到数据库中： 可以自定义插件，保存的时候，即时出码 -> 调用接口将出码结果保存到数据库中。
2. 打开页面从数据库中读取显示，预览的页面数据直接从接口取数据即可。

已打开的页面 keepalive 具体是意思是？

**Reply by stwflyfox** [reply]
您说的是在设计器里的操作，我想实现的是在项目运行时，像阿里低代码有两种运行时，使用schema渲染运行或使用出码运行，使用出码运行增强了性能损失了灵活性，使用schema渲染运行则相反，不同的页面可以根据应用场景使用不同的运行时。有一种应用场景是由我们设计好初始页面，然后由客户自己调整组件位置及显示隠藏等个性化操作，不同客户会有自己的个性化页面副本，然后这个页面在运行时会按照客户读取不同的个性化页面副本做渲染。而且客户随时都可以调整自己的页面副本，即时看到调整后结果，没法统一时间做发布出码打包到项目里去。而tinyEngine没有shcema渲染运行功能，我能想到的就是把客户个性化的页面副本保存时都出码到数据库中，然后在项目里用repl显示对应的页面了。不知道是否有更好的实现方法呢？运行时是项目能否落地的重要环节，希望tinyEngine能有多种出码渲染解决方案，比如仿照repl开发一个适用于tinyEngie的出码渲染，还是用ssr实现渲染等方式..

**Reply by chilingling** [reply]
建议：
客户调整及时看到结果使用 vue-repl 进行预览，调整确认完毕之后有确认发布到生产环境的按钮。然后执行 schema 出码 -> 构建项目 -> 部署到测试环境 -> 部署到生产环境的动作。

即使具备了根据 schema 直接渲染页面的运行时，其实也不建议直接拖拉拽然后保存立刻在生产环境生效，因为这样可能有误操作，造成生产环境的不稳定。
@stwflyfox

**Reply by stwflyfox** [reply]
因为我们开发的是SASS应用，同一个页面不同的客户会有个性化副本，无法走重新构建和发布的流程。由于以前用的表单设计器是渲染schema也就无需重新构建发布。现在只有出码，那就只能在运行时渲染不同的出码副本， 而不是把所有出码副本构建到项目中，出码副本随时会变动的，所以无法走重新构建和发布的流程。

**Reply by xiaofeihe** [reply]
> 因为我们开发的是SASS应用，同一个页面不同的客户会有个性化副本，无法走重新构建和发布的流程。由于以前用的表单设计器是渲染schema也就无需重新构建发布。现在只有出码，那就只能在运行时渲染不同的出码副本， 而不是把所有出码副本构建到项目中，出码副本随时会变动的，所以无法走重新构建和发布的流程。

我们也有同样的使用场景， 也在找 类是方案。

**Reply by stwflyfox** [reply]
目前的解决方法是把页面上会变化的部份做成区块，然后按照条件显示对应的区块。

**Reply by i3web** [reply]
> 您说的是在设计器里的操作，我想实现的是在项目运行时，像阿里低代码有两种运行时，使用schema渲染运行或使用出码运行，使用出码运行增强了性能损失了灵活性，使用schema渲染运行则相反，不同的页面可以根据应用场景使用不同的运行时。有一种应用场景是由我们设计好初始页面，然后由客户自己调整组件位置及显示隠藏等个性化操作，不同客户会有自己的个性化页面副本，然后这个页面在运行时会按照客户读取不同的个性化页面副本做渲染。而且客户随时都可以调整自己的页面副本，即时看到调整后结果，没法统一时间做发布出码打包到项目里去。而tinyEngine没有shcema渲染运行功能，我能想到的就是把客户个性化的页面副本保存时都出码到数据库中，然后在项目里用repl显示对应的页面了。不知道是否有更好的实现方法呢？运行时是项目能否落地的重要环节，希望tinyEngine能有多种出码渲染解决方案，比如仿照repl开发一个适用于tinyEngie的出码渲染，还是用ssr实现渲染等方式..

https://github.com/opentiny/tiny-engine/tree/ospp-2025/runtime-rendering
这个分支已经支持了runtime-render了，但是没合入到主分支。
https://github.com/opentiny/tiny-engine/pull/1661

**Comment by i3web** [comment]
我们也需要一个运行时的渲染引擎。不知道您解决了没有
